### PR TITLE
Make provider queries more resilient

### PR DESCRIPTION
**Pull request by mattiecnvr**
### Description

Minor update to retry when we query for a block or txn by hash and don't get an error but also don't find it. Before this change the scraper would crash if a block or txn was reported as unknown by the provider. This should make it a little less likely that we crash.

### Drive-by changes

No

### Backward compatibility

_Are these changes backward compatible?_

Yes

_Are there any infrastructure implications, e.g. changes that would prohibit deploying older commits using this infra tooling?_

None

### Testing

_What kind of testing have these changes undergone?_

Manual

Ran scraper locally and it is still scraping.

### CODE REVIEW

**Comment by mattiecnvr** [comment]
Flexible on constants, just wanted to get something simple ready.

**Comment by asaj** [comment]
Do we understand *why* we were getting `None` in the first place?

**Comment by mattiecnvr** [comment]
> Do we understand _why_ we were getting `None` in the first place?

Kinda mentioned this in standup but to save this for the record:
We receive `None` when a provider says that the object does not exist. This would happen either if we provide a hash from the wrong blockchain or (in this case) if the provider is behind.

Querying for something which does not exist is not an error for this query, it is just `None`. The underlying `eth_getBlockByHash` or `eth_getTransactionByHash` functions return `null` so this is to spec.

What happened specifically that lead to this PR was that bsctestnet was configured with 2 nodes, one that was frequently 5xx-ing and another which was > 1k blocks behind. The consequence of this was that a quorum would be reached with one error and one valid null response, and then we would report that the block was not found and crash the program.

In this specific part of the code, we know that we are querying a block we found so not finding it implies there is actually a need to retry.

File: rust/chains/abacus-ethereum/src/provider.rs [hunk 1]
```diff
@@ -110,3 +106,25 @@ impl MakeableWithProvider for AbacusProviderBuilder {
...
+/// object we are looking for yet.
+async fn get_with_retry_on_none<T, F, O, E>(get: F) -> eyre::Result<T>
+where
+    F: Fn() -> O,
+    O: Future<Output = Result<Option<T>, E>>,
+    E: std::error::Error + Send + Sync + 'static,
+{
+    for i in 0..3 {
+        if i > 0 {
+            sleep(Duration::from_secs(5)).await;
```
Comment (tkporter) [inline rust/chains/abacus-ethereum/src/provider.rs:121]: any reason to not put this before the `continue`? I figure if it fails once then it's unlikely to immediately work if retried again, so sleeping even on the first failure is beneficial - though maybe there's reason to believe that intuition is wrong

**Review by tkporter — COMMENTED** [review]
Has this ever happened to us in a more "normal" case apart from the bsctestnet example you gave? I.e. when providers seem to be having totally normally but we just happen to be making these requests far too quickly

**Comment by mattiecnvr** [comment]
> Has this ever happened to us in a more "normal" case apart from the bsctestnet example you gave? I.e. when providers seem to be having totally normally but we just happen to be making these requests far too quickly

Here, I went ahead and made a query to spot all the times it has happened. In the past 10 days there were 110 restarts that this PR could have mitigated. Since the start of this month there were 
https://cloudlogging.app.goo.gl/J2LRKxPukTZnmFCG7

If you dig around you will see it happens with other providers as well.

File: rust/chains/abacus-ethereum/src/provider.rs [hunk 2]
```diff
@@ -41,27 +45,19 @@ where
...
 {
     async fn get_block_by_hash(&self, hash: &H256) -> eyre::Result<BlockInfo> {
-        let block = self
-            .provider
-            .get_block(*hash)
-            .await?
-            .ok_or_else(|| eyre!("Could not find block with hash {}", hash))?;
-        Ok(BlockInfo {
+        let block = get_with_retry_on_none(|| self.provider.get_block(*hash)).await?;
+        return Ok(BlockInfo {
```
Comment (tkporter) [inline rust/chains/abacus-ethereum/src/provider.rs:49]: any reason for adding the `return`?

File: rust/chains/abacus-ethereum/src/provider.rs (hunk 1 above)
Comment (tkporter) [reply rust/chains/abacus-ethereum/src/provider.rs:121]: still curious how you feel here. not passionate though

**Review by tkporter — APPROVED** [review]
Agree it's worth the fix!

File: rust/chains/abacus-ethereum/src/provider.rs (hunk 1 above)
Comment (mattiecnvr) [reply rust/chains/abacus-ethereum/src/provider.rs]: Yeah, that makes sense, changing

File: rust/chains/abacus-ethereum/src/provider.rs (hunk 2 above)
Comment (mattiecnvr) [reply rust/chains/abacus-ethereum/src/provider.rs]: Not really. Fixed

**Pull request merged**